pkg/patterns/declarative/pkg/watch/dynamic.go

```diff
@@ -50,32 +54,51 @@ func (dw *dynamicWatch) newDynamicClient(gvk schema.GroupVersionKind) (dynamic.R
 
 // Add registers a watch for changes to 'trigger' filtered by 'options' to raise an event on 'target'
 func (dw *dynamicWatch) Add(trigger schema.GroupVersionKind, options metav1.ListOptions, target metav1.ObjectMeta) error {
-	log := log.Log
-
 	client, err := dw.newDynamicClient(trigger)
 	if err != nil {
 		return fmt.Errorf("creating client for (%s): %v", trigger.String(), err)
 	}
 
+	go func() {
+		lastResourceVersion := ""
+		for {
+			lastResourceVersion = dw.watchUntilClosed(lastResourceVersion, client, trigger, options, target)
+
+			time.Sleep(WatchDelay)
+		}
+	}()
+
+	return nil
+}
+
+func (dw *dynamicWatch) watchUntilClosed(lastResourceVersion string, client dynamic.ResourceInterface, trigger schema.GroupVersionKind, options metav1.ListOptions, target metav1.ObjectMeta) string {
+	log := log.Log
+
+	options.ResourceVersion = lastResourceVersion
```

Comment:
I'm wondering about what happens if lastResourceVersion is too old...

I was thinking there must be a function in our libraries for it - and I find this...
https://github.com/kubernetes/client-go/blob/bad694ec20e3ad1bf3ede898519b4948962639b7/tools/watch/retrywatcher.go

But it looks like that one doesn't handle a "resource version too old" error either... @liggitt should we worry about this scenario?

Reply:
that one explicitly doesn't handle a resourceversiontoo old scenario because it has no way to re-establish a list:

https://github.com/kubernetes/client-go/blob/bad694ec20e3ad1bf3ede898519b4948962639b7/tools/watch/retrywatcher.go#L48-L49

there is a helper that does handle that now: https://github.com/kubernetes/client-go/blob/bad694ec20e3ad1bf3ede898519b4948962639b7/tools/watch/informerwatcher.go#L59-L62

Reply:
Thank you! 

It seems that we're better off not managing `ResourceVersion` for now. If we just resume the watch on a failure or drop then we will always reconcile when we need too. We will over-reconcile.

Reply:
Thank you!

It seems that we're better off not managing `ResourceVersion` for now. If we just resume the watch on a failure or drop then we will always reconcile when we need too. We will over-reconcile.

 WDYT @liggitt ?

Reply:
> If we just resume the watch on a failure or drop then we will always reconcile when we need too. We will over-reconcile.

not having looked through the code, what method would you use to resume the watch without managing a resource version?

Reply:
It's not really resuming now, it's just restarting with an empty resource version, basically `client.Watch(metav1.ListOptions{})`

Reply:
I got the idea from your comment here: https://github.com/kubernetes/kubernetes/issues/54878#issuecomment-357575276

Reply:
it is more typical to use an informer in cases like this

Reply:
We did think the optimization here could be to only reset resourceVersion when we observe the "too old" message.  I'm guessing that looks like this: https://github.com/kubernetes/client-go/blob/191b4b1228013836d532375e8cf85986be65edfb/tools/watch/retrywatcher.go#L204

These are going to be watches of the handful of objects each addon-operator created, hopefully tightly filtered by labels.  The reconciliation might be relatively expensive however because we don't (yet) have a way to skip doing an apply - we have to run it speculatively.

I don't love that an informer would buffer the objects in memory, but there won't be many of them, and it could save us from a spurious apply.

And I'm guessing we do now have an example of dynamic informers:

https://github.com/kubernetes/client-go/blob/master/dynamic/dynamicinformer/informer_test.go#L128

Reply:
I filed #39 to better handle this